tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,182 @@
+# Stateful tests for arbitrary Zarr stores.
+
+import asyncio
+
+import hypothesis.strategies as st
+from hypothesis import note
+from hypothesis.stateful import (
+    RuleBasedStateMachine,
+    invariant,
+    precondition,
+    rule,
+)
+
+from zarr.abc.store import Store as StoreABC
+from zarr.buffer import Buffer, default_buffer_prototype
+from zarr.store import LocalStore, MemoryStore, RemoteStore
+
+# from strategies_store import StoreStatefulStrategies, key_ranges
+from zarr.testing.strategies import key_ranges, paths
+
+# zarr spec: https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html
+
+
+class SyncStoreWrapper():
+    def __init__(self, store):
+        """Store to hold async functions that map to StoreABC abstract methods"""
+        self.store = store
+        # Unfortunately, hypothesis' stateful testing infra does not support asyncio
+
+    # So we redefine sync versions of the Store API.
+    # https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+    def set(self, key, data_buffer):  # buffer is value
+        return asyncio.run(self.store.set(key, data_buffer))
+
+    async def list(self):
+        paths = [path async for path in self.store.list()]
+        # note(f'(store set) store {paths=}, type {type(paths)=}, {len(paths)=}')
+        return paths
+
+    async def get(
+        self,
+        key,
+    ):
+        obs = await self.store.get(key, prototype=default_buffer_prototype())
+        return obs
+
+    async def get_partial_values(self, key_ranges):
+        obs_maybe = await self.store.get_partial_values(
+            prototype=default_buffer_prototype(), key_ranges=key_ranges
+        )
+        return obs_maybe
+
+    async def delete(self, path):  # path is key
+        await self.store.delete(path)
+
+    async def empty(self):
+        await self.store.empty()
+
+    async def clear(self):
+        await self.store.clear()
+
+    async def exists(self, key):
+        raise NotImplementedError
+
+    async def list_dir(self, prefix):
+        raise NotImplementedError
+
+    async def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    async def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    async def supports_listing(self):
+        raise NotImplementedError
+
+    async def supports_partial_writes(self):
+        raise NotImplementedError
+
+    async def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    def __init__(self):  # look into using run_machine_as_test()
+        super().__init__()
+        self.model = {}
```

Comment:
```suggestion
        self.model: dict[str, bytes] = {}
```

Reply:
missed this earlier but addressed in https://github.com/zarr-developers/zarr-python/pull/2070/commits/8e6f03c02c920e4979d70fb154dbb2a33f4ca1ee